### ISSUE TITLE
Crash using "terraform provider mirror" when using invalid syntax

### ISSUE THREAD

**Issue by billyfoss** [hashicorp/terraform#29087]
When I defined a provider key twice, the terraform provider mirror command will crash in 0.15 and 1.0.1.  In 0.13 and 0.14, an appropriate error is reported.  

### Terraform Version
```
Terraform v1.0.1
on linux_amd64
```
Note: also crashes on 0.15.5, but it provides a helpful error message on 0.14.7 and 0.13.7

### Terraform Configuration Files

```terraform
# Define Terraform providers
terraform {
  required_providers {
    time = {   # NOTE: I copied and pasted, but failed to update the key.
      source  = "hashicorp/template"
      version = "~> 2.2"
    }
    time = {
      source  = "hashicorp/time"
      version = "~> 0.6"
    }
  }
}
```

### Crash Output

Please let me know if you cannot recreate this with the sample file.

### Expected Behavior

In 0.14.7, we see 
```
Error: Attribute redefined

  on versions.tf line 8, in terraform:
   8:     time = {

The argument "time" was already set at versions.tf:4,5-9. Each argument may be
set only once.
```

### Actual Behavior

In 0.15.5 and 1.0.1, we see 

```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0x9e58ba]

goroutine 1 [running]:
github.com/hashicorp/terraform/internal/configs.(*Config).addProviderRequirements(0x0, 0xc00041ee00, 0x1, 0xc00041eb00, 0x1, 0x1)
        /home/circleci/project/project/internal/configs/config.go:241 +0x3a
github.com/hashicorp/terraform/internal/configs.(*Config).ProviderRequirements(...)
        /home/circleci/project/project/internal/configs/config.go:188
github.com/hashicorp/terraform/internal/command.(*ProvidersMirrorCommand).Run(0xc00002c000, 0xc00004c0b0, 0x1, 0x1, 0x0)
        /home/circleci/project/project/internal/command/providers_mirror.go:76 +0x5bb
github.com/mitchellh/cli.(*CLI).Run(0xc00001a500, 0xc0007383f0, 0x2848e95, 0x10)
        /home/circleci/go/pkg/mod/github.com/mitchellh/cli@v1.1.2/cli.go:262 +0x41a
main.wrappedMain(0x0)
        /home/circleci/project/project/main.go:384 +0x1112
main.realMain(0x0)
        /home/circleci/project/project/main.go:115 +0x554
main.main()
        /home/circleci/project/project/main.go:58 +0x25




!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!

Terraform crashed! This is always indicative of a bug within Terraform.
A crash log has been placed at "crash.508489984.log" relative to your current
working directory. It would be immensely helpful if you could please
report the crash with Terraform[1] so that we can fix this.

When reporting bugs, please include your terraform version. That
information is available on the first line of crash.log. You can also
get it by running 'terraform --version' on the command line.

SECURITY WARNING: the "crash.508489984.log" file that was created may contain
sensitive information that must be redacted before it is safe to share
on the issue tracker.

[1]: https://github.com/hashicorp/terraform/issues

!!!!!!!!!!!!!!!!!!!!!!!!!!! TERRAFORM CRASH !!!!!!!!!!!!!!!!!!!!!!!!!!!!
```

### Steps to Reproduce

1. terraform providers mirror /tmp/.terraform.d/plugins

**Comment by apparentlymart** [comment]
Hi @billyfoss! Sorry for the long silence here.

I tried to reproduce this with v1.3.0 today and found it now works again as originally described, with the error message about a duplicate argument.

This particular message comes from the underlying HCL library and is reporting the general rule that argument names within a particular block must be unique. Therefore I assume what was happening in v0.15 and v1.1 is that some subsequent code was trying to gracefully recover from the error and produce a partial result in spite of it, but it wasn't being careful enough in its analysis of the incomplete data structure and therefore inadvertently crashed.

I haven't been able to track down exactly which subsequent change fixed this, but I do remember us doing some general robustness improvements to the error handling and recovery in the configuration loader in a recent release and so I expect we unknowingly fixed that bug as part of that, either by being more cautious about the incomplete/invalid parts of the configuration data structure or just bailing out earlier and returning the errors learned so far instead of trying to do further analysis downstream.

Since this now seems to work again in recent versions of Terraform, I'm going to close this issue. If you or someone else is seeing a similar problem in Terraform v1.3.0 or later then I'd ask that you please open a new issue to cover it so that we can capture a fresh set of debug information and stack trace that will correlate better with the changes we made to the config decoder that seem to have fixed this bug.

Thanks for reporting this!
